src/tools/illink/src/ILLink.Shared/DataFlow/ForwardDataFlowAnalysis.cs

```diff
@@ -265,9 +265,23 @@ public void Fixpoint (TControlFlowGraph cfg, TLattice lattice, TTransfer transfe
 						// Using predecessors in the finally. But not from outside the finally.
 						exceptionFinallyState = lattice.Top;
 						foreach (var predecessor in cfg.GetPredecessors (block)) {
-							var isPredecessorInFinally = cfgState.TryGetExceptionFinallyState (predecessor.Block, out TValue predecessorFinallyState);
+							var isPredecessorInFinally = cfgState.TryGetExceptionFinallyState (predecessor.Block, out TValue predecessorState);
 							Debug.Assert (isPredecessorInFinally);
-							exceptionFinallyState = lattice.Meet (exceptionFinallyState.Value, predecessorFinallyState);
+
+							// Propagate state through all finally blocks.
+							foreach (var exitedFinally in predecessor.FinallyRegions) {
+								TValue oldFinallyInputState = cfgState.GetFinallyInputState (exitedFinally);
+								TValue finallyInputState = lattice.Meet (oldFinallyInputState, predecessorState);
+
+								cfgState.SetFinallyInputState (exitedFinally, finallyInputState);
+								if (!changed && !finallyInputState.Equals (oldFinallyInputState))
+									changed = true;
+
+								TBlock lastFinallyBlock = cfg.LastBlock (exitedFinally);
+								predecessorState = cfgState.Get (lastFinallyBlock).Current;
+							}
```

Comment:
Just curious if I understand this correctly:
```C#
try
{
}
finally // OuterFinally
{
    try
    {
    }
    finally // InnerFinally
    {
    }

    {  // BlockOfInterest
    }
}
```

The goal is to propagate the finally exception state from InnerFinally to BlockOfInterest, right?

Can we dedupe this with the code on line 222? It does exactly the same thing for the "normal" state. It makes sense that we have to do it twice (once for the normal state and once for the finally exception state), but then it would make sense to have one piece of code to do it for both - it would make it easy to see what's going on.